### PR TITLE
avm2: partially implement Graphics.drawTriangles and drawGraphicsData

### PR DESCRIPTION
**Pull request by SuchAFuriousDeath**
Progresses #9017 (even seems to make it playable, although the stone texture still looks weird, see the comment I made)
UV mapping, positive and negative culling are unimplemented.

### CODE REVIEW

**Comment by n0samu** [comment]
Welcome, and thank you for contributing! 😃

I'm thinking #9596 could potentially be fixed by this as well!

**Comment by torokati44** [comment]
> I'm thinking https://github.com/ruffle-rs/ruffle/issues/9596 could potentially be fixed by this as well!

Almost, but without UV mapping, the textures don't distort together with the mesh. The mesh almost looks like a mask applied to the textures as backgrounds.

**Comment by n0samu** [comment]
Ah okay, it's a start!

**Comment by n0samu** [comment]
@SuchAFuriousDeath Can you run `cargo fmt` to fix the test failures?

**Comment by SuchAFuriousDeath** [comment]
Here you go

File: core/src/avm2/globals/flash/display/graphics.rs [hunk 1]
```diff
@@ -865,8 +865,278 @@ pub fn draw_path<'gc>(
...
+pub fn draw_triangles<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    this: Object<'gc>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    if let Some(this) = this.as_display_object() {
+        if let Some(mut drawing) = this.as_drawing(activation.context.gc_context) {
+            let vertices = args.get_object(activation, 0, "vertices")?;
+
+            let indices = args.get_object(activation, 1, "indices").ok();
```
Comment (Dinnerbone) [inline core/src/avm2/globals/flash/display/graphics.rs:895]:
```suggestion
            let indices = args.try_get_object(activation, 1);
```

File: core/src/avm2/globals/flash/display/graphics.rs [hunk 2]
```diff
@@ -865,8 +865,278 @@ pub fn draw_path<'gc>(
...
+
+            let uvt_data = args.get_object(activation, 2, "uvtData").ok();
```
Comment (Dinnerbone) [inline core/src/avm2/globals/flash/display/graphics.rs:897]:
```suggestion
            let uvt_data = args.try_get_object(activation, 2);
```

File: core/src/avm2/globals/flash/display/graphics.rs [hunk 3]
```diff
@@ -942,95 +1212,410 @@ pub fn draw_path<'gc>(
...
-    Ok(Value::Undefined)
+fn culling_to_triangle_culling<'gc>(culling: AvmString) -> Result<TriangleCulling, Error<'gc>> {
+    if &culling == b"none" {
+        Ok(TriangleCulling::None)
+    } else if &culling == b"positive" {
+        Ok(TriangleCulling::Positive)
+    } else if &culling == b"negative" {
+        Ok(TriangleCulling::Negative)
+    } else {
+        Err(format!("Unknown culling {:?}", culling).into())
```
Comment (Dinnerbone) [inline core/src/avm2/globals/flash/display/graphics.rs:1337]:
```suggestion
        Err(make_error_2008(activation, "culling")
```

**Review by Dinnerbone — CHANGES_REQUESTED** [review]
Few minor changes. Thank you very much!

**Comment by SuchAFuriousDeath** [comment]
#9829 looks playable now

**Comment by SuchAFuriousDeath** [comment]
By the way, just wondering, is there any reason why graphicsshaderfill is not in activation.avm2().classes()? 
https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/display/GraphicsShaderFill.html

**Comment by Dinnerbone** [comment]
> By the way, just wondering, is there any reason why graphicsshaderfill is not in activation.avm2().classes()? https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/display/GraphicsShaderFill.html

Simply because nobody's needed it anywhere, so nobody added it yet.

**Comment by SuchAFuriousDeath** [comment]
#12022 requires negative culling to be implemented, but unfortunately requires a lot of other functions that are currently stubbed

**Review by Dinnerbone — APPROVED** [review]
Thank you!

**Pull request merged**